File: src/adapters/base/PaymentAdapter.ts
import {
  CreateTransactionParams,
  Transaction,
  TransactionResult,
  RefundParams,
  Refund,
  TransactionStatus,
} from '../../types';

// Interfaz que deben implementar todos los adaptadores de pago
export interface PaymentAdapter {
  // Nombre del proveedor
  readonly providerName: string;

  // Crear una nueva transacción
  createTransaction(params: CreateTransactionParams): Promise<Transaction>;

  // Confirmar una transacción (cuando el usuario vuelve del pago)
  confirmTransaction(token: string): Promise<TransactionResult>;

  // Consultar el estado de una transacción
  getTransactionStatus(transactionId: string): Promise<TransactionStatus>;

  // Reembolsar una transacción
  refundTransaction(params: RefundParams): Promise<Refund>;

  // Validar webhook (notificaciones del proveedor)
  validateWebhook(payload: any, signature?: string): boolean;
}

File: kotlin/src/main/kotlin/cl/payments/adapters/base/PaymentAdapter.kt
package cl.payments.adapters.base

import cl.payments.types.*

// Interfaz que deben implementar todos los adaptadores de pago
interface PaymentAdapter {
    // Nombre del proveedor
    val providerName: String

    // Crear una nueva transacción
    suspend fun createTransaction(params: CreateTransactionParams): Transaction

    // Confirmar una transacción (cuando el usuario vuelve del pago)
    suspend fun confirmTransaction(token: String): TransactionResult

    // Consultar el estado de una transacción
    suspend fun getTransactionStatus(transactionId: String): TransactionStatus

    // Reembolsar una transacción
    suspend fun refundTransaction(params: RefundParams): Refund

    // Validar webhook (notificaciones del proveedor)
    fun validateWebhook(payload: Map<String, Any>, signature: String?): Boolean
}